Return position `impl Trait` in traits (#3425)
rust-lang/rfcs: text/0000-return-position-impl-trait-in-traits.md

---
tags: draft-rfc
title: Return position `impl Trait` in traits
---

- Feature Name: return_position_impl_trait_in_traits
- Start Date: 2023-04-27
- RFC PR: [rust-lang/rfcs#3193](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
- Initiative: [impl trait initiative](https://github.com/rust-lang/impl-trait-initiative)

# Summary
[summary]: #summary

* Permit `impl Trait` in fn return position within traits and trait impls.

> * Allow `async fn` in traits and trait impls to be used interchangeably with its equivalent `impl Trait` desugaring.

Reviewer: One consequence of this is that it basically stabilizes the `impl Future` desugaring. I don't imagine we'll want to change it, and enough things depend on the `impl Future` desugaring now that we probably can't change it practically already, but it still means we're committing to what in some ways feel like internal implementation details to me.

* Allow trait impls to `#[refine]` an `impl Trait` return type with added bounds or a concrete type.[^refine]

# Motivation
[motivation]: #motivation

The `impl Trait` syntax is currently accepted in a variety of places within the Rust language to mean "some type that implements `Trait`" (for an overview, see the [explainer] from the impl trait initiative). For function arguments, `impl Trait` is [equivalent to a generic parameter][apit] and it is accepted in all kinds of functions (free functions, inherent impls, traits, and trait impls).

In return position, `impl Trait` [corresponds to an opaque type whose value is inferred][rpit]. In that role, it is currently accepted only in free functions and inherent impls. This RFC extends the support for return position `impl Trait` in functions in traits and trait impls.

[explainer]: https://rust-lang.github.io/impl-trait-initiative/explainer.html
[apit]: https://rust-lang.github.io/impl-trait-initiative/explainer/apit.html
[rpit]: https://rust-lang.github.io/impl-trait-initiative/explainer/rpit.html

## Example use case

> The use case for `-> impl Trait` in trait functions is similar to its use in other contexts: traits often wish to return "some type" without specifying the exact type. As a simple example that we will use through the RFC, consider the `NewIntoIterator` trait, which is a variant of the existing `IntoIterator` that uses `impl Iterator` as the return type:

Reviewer: This makes it sound like the primary benefit of RPITIT is ergonomics, rather than adding significant new expressiveness to the language. In other words, it makes some things easier to say but doesn't add the ability to say new things. Is that right?

Reviewer 2: Yes and no. For example, you can do this today:
```rust
trait Foo {
    type Fut<'a>: std::future::Future<Output = ()> + 'a
    where
        Self: 'a;
        
    fn fut(&self) -> Self::Fut<'_>;
}

// Foo::fut is essentially an async fn:
async fn with<F: Foo>(foo: &F) {
    foo.fut().await
}
```
But, as in the RFC, this is not dyn-safe:
```rust
// error[E0191]: the value of the associated type `Fut` (from trait `Foo`) must be specified
async fn with_dyn(foo: &dyn Foo) {
    foo.fut().await
}
```
However, *implementing* `Foo` as written above requires specifying the type of `Foo::Fut` (sometimes annoying, sometimes impossible without type-alias impl-trait), despite the fact that such a definition is entirely redundant with the definition of `Foo::fut`.

<!--However, the desugaring to achieve that effect looks somewhat different than other cases of impl trait in return position. This is because we cannot desugar to a type alias in the surrounding module; we need to desugar to an associated type (effectively, a type alias in the trait).-->

Consider the following trait:

```rust
trait IntoIntIterator {
    fn into_int_iter(self) -> impl Iterator<Item = u32>;
}
```

The semantics of this are analogous to introducing a new associated type within the surrounding trait;

```rust
trait IntoIntIterator { // desugared
    type IntoIntIter: Iterator<Item = u32>;
    fn into_int_iter(self) -> Self::IntoIntIter;
}
```

When using `-> impl Trait`, however, there is no associated type that users can name.

By default, the impl for a trait like `IntoIntIterator` must also use `impl Trait` in return position.

```rust
impl IntoIntIterator for Vec<u32> {
    fn into_int_iter(self) -> impl Iterator<Item = u32> {
        self.into_iter()
    }
}
```

> It can, however, give a more specific type with `#[refine]`:[^refine]

Reviewer: Is there a rationale given somewhere for requiring `#[refine]`?

Reviewer 3: https://github.com/rust-lang/rfcs/pull/3245

Author: I think the RFC includes the rationale and is linked to by the footnote, so resolving. Feel free to unresolve if you think there should be more explicit text here.

```rust
impl IntoIntIterator for Vec<u32> {

>     #[refine]

Reviewer 4: Michael has a PR that implement these checks. So resolving this comment ...

    fn into_int_iter(self) -> impl Iterator<Item = u32> + ExactSizeIterator {
        self.into_iter()
    }
    
    // ..or even..

    #[refine]
    fn into_int_iter(self) -> std::vec::IntoIter<u32> {
        self.into_iter()
    }
}
```

Users of this impl are then able to rely on the refined return type, as long as the compiler can prove this impl specifically is being used. Conversely, in this example, code that is generic over the trait can only rely on the fact that the return type implements `Iterator<Item = u32>`.

### async fn desugaring

> `async fn` always desugars to a regular function returning `-> impl Future`. When used in a trait, the `async fn` syntax can be used interchangeably with the equivalent desugaring in the trait and trait impl:

Reviewer: Presumably the reason for doing this is so that we can use `#[refine]` to make stricter promises on the impl? For example, adding a `Send` bound at the impl level.

Author: That's one (another example of an added promise would be that you don't capture `self`). A whole other reason would be to include code that runs at call time instead of during `poll`.

    async fn do_something(&self);
}

// OK!
impl UsesAsyncFn for MyType {
    fn do_something(&self) -> impl Future<Output = ()> + '_ {
        async {}
    }
}
```
```rust
trait UsesDesugaredFn {
    // Equivalent to:
    // async fn do_something(&self);
    fn do_something(&self) -> impl Future<Output = ()> + '_;
}

// Also OK!
impl UsesDesugaredFn for MyType {
    async fn do_something(&self) {}
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## Equivalent desugaring for traits

Each `-> impl Trait` notation appearing in a trait fn return type is effectively desugared to an anonymous associated type. In this RFC, we will use the placeholder name `$` when illustrating desugarings and the like.

As a simple example, consider the following (more complex examples follow):

```rust
trait NewIntoIterator {
    type Item;
    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
}

// becomes

trait NewIntoIterator {
    type Item;

    type $: Iterator<Item = Self::Item>;

    fn into_iter(self) -> <Self as NewIntoIterator>::$;
}
```

```rust
trait SomeTrait {
    fn method<P0, ..., Pm>()
}

> ```

Reviewer: What is this example meant to show?

Author: I think it was left by mistake, removing.

    fn iter<'a>(&'a self) -> impl Iterator<Item = Self:Item<'a>>;
}

// Since 'a is named in the bounds, it is captured.
// `RefIterator` thus becomes:

trait RefIterator for Vec<u32> {
    type Item<'me>
    where 
        Self: 'me;

    type $<'a>: impl Iterator<Item = Self::Item<'a>>
    where 
        Self: 'a; // Implied bound from fn

    fn iter<'a>(&'a self) -> Self::$<'a>;
}
```

## Validity constraint on impls

Given a trait method where `impl Trait` appears in return position,

```rust
trait Trait {
    fn method() -> impl T_0 + ... + T_m;
}
```

where `T_0 + ... + T_m` are bounds, for any impl of that trait to be valid, the following conditions must hold:

* The return type named in the corresponding impl method must implement all bounds `T_0 + ... + T_m` specified in the trait.
* Either the impl method must have `#[refine]`,[^refine] OR
    * The impl must use `impl Trait` syntax to name the corresponding type, and
    * The return type in the trait must implement all bounds `I_0 + ... + I_n` specified in the impl return type. (Taken with the first outer bullet point, we can say that the bounds in the trait and the bounds in the impl imply each other.)

> [^refine]: `#[refine]` was added in [RFC 3245: Refined trait implementations](https://rust-lang.github.io/rfcs/3245-refined-impls.html). This feature is not yet stable.

Reviewer: Does this mean this RFC is dependent on 3245 being stabilized first, or is the plan to include a subset of 3245 in this RFC?

Author: Added clarifying text to this footnote; it just means any code that requires `#[refine]` won't work until that feature is stabilized. Simple examples without refinement will continue to work.

Additionally, using `-> impl Trait` notation in an impl is only legal if the trait also uses that notation.

```rust
trait NewIntoIterator {
    type Item;
    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
}

// OK:
impl NewIntoIterator for Vec<u32> {
    type Item = u32;
    fn into_iter(self) -> impl Iterator<Item = u32> {
        self.into_iter()
    }
}

// OK:
impl NewIntoIterator for Vec<u32> {
    type Item = u32;
    #[refine]
    fn into_iter(self) -> impl Iterator<Item = u32> + DoubleEndedIterator {
        self.into_iter()
    }
}

// OK:
impl NewIntoIterator for Vec<u32> {
    type Item = u32;
    #[refine]
    fn into_iter(self) -> std::vec::IntoIter<u32> {
        self.into_iter()
    }
}

// Not OK:
impl NewIntoIterator for Vec<u32> {
    type Item = u32;
    fn into_iter(self) -> std::vec::IntoIter<u32> {
        self.into_iter()
    }
}

> ```

Reviewer: This example is Not OK because it's missing the `#[refine]` attribute, right?

Author: Right. Added to the comment.

### Interaction with `async fn` in trait

This RFC modifies the “Static async fn in traits” RFC so that async fn in traits may be satisfied by implementations that return `impl Future<Output = ...>` as long as the return-position impl trait type matches the async fn's desugared impl trait with the same rules as above.

```rust
trait Trait {
  async fn async_fn();
  
  async fn async_fn_refined();
}

impl Trait for MyType {
  fn async_fn() -> impl Future<Output = ()> + '_ { .. }
  
  #[refine]
  fn async_fn_refined() -> BoxFuture<'_, ()> { .. }
}
```

Similarly, the equivalent `-> impl Future` signature in a trait can be satisfied by using `async fn` in an impl of that trait.

> ## Nested impl traits

Reviewer 3: So, this example compiles today:

```rust
#![feature(return_position_impl_trait_in_trait)]

trait Foo {
    fn bar(&self) -> (impl std::fmt::Debug, impl std::fmt::Debug);
}

impl Foo for u32 {
    fn bar(&self) -> (impl std::fmt::Debug, impl std::fmt::Debug) {
        (22, 44)
    }
}
```

this makes sense, given the desugaring described in the RFC, but I didn't find any explicit mention of it (though upon re-reading I realize now it's implicit in sentences like "Each `-> impl Trait` notation appearing in a trait fn return type is effectively desugared to an anonymous associated type. "). We should add an explicit section on this, or perhaps generalize this one to "legal positions for `impl Trait` to appear".

Author: I think the intent should be that anything that is legal in a free function should be legal here, so I'll update the section to say that.

Author: @nikomatsakis Can you check that the updated text resolves your concerns?

[gankra]: https://github.com/rust-lang/rfcs/pull/3193#issuecomment-965505149

```rust
fn is_palindrome<Iter, T>(iterable: Iter) -> bool
where
    Iter: IntoIterator<Item = T>,
    Iter::IntoIter: DoubleEndedIterator,
    T: Eq;
```

The `NewIntoIterator` trait used as an example in this RFC, however, doesn't support this kind of usage, because there is no way for users to name the `IntoIter` type (and, as discussed in the previous section, there is no way for users to migrate to a named associated type, either!). The same problem applies to async functions in traits, which sometimes wish to be able to [add `Send` bounds to the resulting futures](https://rust-lang.github.io/async-fundamentals-initiative/evaluation/challenges/bounding_futures.html).

The [future possibilities](#future-possibilities) section discusses a planned extension to support naming the type returned by an impl trait, which could work to overcome this limitation for clients.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

### Does auto trait leakage still occur for `-> impl Trait` in traits?

> Yes, so long as the compiler has enough type information to figure out which impl you are using. In other words, given a trait function `SomeTrait::foo`, if you invoke a function `<T as SomeTrait>::foo()` where the self type is some generic parameter `T`, then the compiler doesn't really know what impl is being used, so no auto trait leakage can occur. But if you were to invoke `<u32 as SomeTrait>::foo()`, then the compiler could resolve to a specific impl, and hence a specific [impl trait type alias][tait], and auto trait leakage would occur as normal.

Reviewer: Once we have associated return types / return type notation, I wonder if we can change this and if it'd be worth changing it? As long as it's not too burdensome to be explicit, that seems like it might be a better situation than the current state of things.

Author: It _might_ be worth changing, but not as part of this RFC. If we ever remove auto trait leakage I think we should do it globally, not as part of individual new expansions of `-> impl Trait`, and this would of course require an edition migration.

We also aren't that close to removing it; we would need some kind of syntax that means "this type impls `Send` if `T: Send`" for some input type `T`, which we don't have, and that was a key part of the reason for auto trait leakage in the first place.

[tait]: https://rust-lang.github.io/impl-trait-initiative/explainer/tait.html

### Can traits migrate from a named associated type to `impl Trait`?

> Not compatibly, no, because they would no longer have a named associated type.

Reviewer 3: ```suggestion
Not compatibly, no, because they would no longer have a named associated type. The "future directions" section discusses the possibility of allowing users to explicitly give a name for the associated type created, which would enable this use case.
```

### Can traits migrate from `impl Trait` to a named associated type?

> Generally yes, but all impls would have to be rewritten.

Reviewer: We'd probably need TAIT too, right?

Reviewer 3: Good callout.

Reviewer 3: I think it'd be worth expanding this answer to go into a bit more detail about *why*

Reviewer 3: ```suggestion
Generally yes, but all impls would have to be rewritten to include the definition of the associated type. In many cases, some form of type-alias impl trait (or impl trait in associated type values) would also be required. For example, if we changed the `IntoIntIterator` trait from the motivation to use an explicit associated type..

```rust
trait IntoIntIterator {
    type IntIter: Iterator<Item = u32>;
    fn into_iter(self) -> Self::IntIter;
}
```

...then impls like...

```rust
impl IntoIntIterator for MyType {
    fn into_int_iter(self) -> impl Iterator<Item = u32> {
        (0..self.len()).map(|x| x * 2)
    }
}
```

...would no longer compile, because they are not specifying the value of the `IntIter` associated type. Moreover, the value for this type would be impossible to express without `impl Trait` notation, as it embeds a closure type.

### Would there be any way to make it possible to migrate from `impl Trait` to a named associated type compatibly?

> Potentially! There have been proposals to allow the values of associated types that appear in function return types to be inferred from the function declaration. So the trait has `fn method(&self) -> Self::Iter` and the impl has `fn method(&self) -> impl Iterator`, then the impl would also be inferred to have `type Iter = impl Iterator` (and the return type rewritten to reference it). This may be a good idea, but it is not proposed as part of this RFC.

Reviewer 3: ```suggestion
Potentially! There have been proposals to allow the values of associated types that appear in function return types to be inferred from the function declaration. So, using the example from the previous question, the impl for `IntoIntIterator` could infer the value of `IntIter` based on the return type of `into_int_iter`. This may be a good idea, but it is not proposed as part of this RFC.
```

Author: This will be an incomplete solution once we support dyn dispatch because you'd have to avoid breaking uses of `dyn Trait` that don't specify the associated type. There's also a future possibilities section on this that needs updating.

   ```rust
   fn iterate_over_anything_as_long_as_it_is_vec<T>(
       vec: impl IntoIterator<IntoIter = std::vec::IntoIter<T>, Item = T>
   )
   ```
1. Bounding the iterator by adding extra bounds like `DoubleEndedIterator` is useful, but not the common case for `IntoIterator`. It therefore shouldn't be overly burdensome to use a (reasonably ergonomic) special syntax in the cases where it's needed.
1. Using `IntoIterator` with dynamic dispatch would be surprising; more common would be to call `.into_iter()` using static dispatch and then pass the resulting iterator to a function that uses dynamic dispatch.
1. If we did use `IntoIterator` with dynamic dispatch, the resulting iterator being dynamically dispatched would make the most sense.

Therefore, if we were writing `IntoIterator` today, it would probably use `-> impl Trait` in return position instead of having an explicit `IntoIter` type.

The same is not true for `Iterator::Item`: because `Item` is so central to what an `Iterator` is, and because it rarely makes sense to use an opaque type for the item, it would remain an explicit associated type.

> # Prior art

Reviewer 3: the RFC should address the points raised by @lcnr 

```suggestion
## Should library traits migrate to use `impl Trait`?

Potentially, but not necessarily. Using `impl Trait` in traits imposes some limitations on generic code referencing those traits. While `impl Trait `desugars internally to an associated type, that associated type is anonymous and cannot be directly referenced by users, which prevents them from putting bounds on it or naming it for use in struct declarations. This is similar to `-> impl Trait` in free and inherent functions, which also returns an anonymous type that cannot be directly named. Just as in those cases, this likely means that widely used libraries should avoid the use of `-> impl Trait` and prefer to use an explicit named associated type, at least until some of the "future possibilities" are completed. However, this decision is best reached on a case-by-case basis: the real question is whether the bounds named in the trait will be sufficient, or whether users will wish to put additional bounds. In a trait like `IntoIterator`, for example, it is common to wish to bound the resulting iterator with additional traits, like `ExactLenIterator`. But given a trait that returns `-> impl Debug`, this concern may not apply.
```

[prior-art]: #prior-art

There are a number of crates that do desugaring like this manually or with procedural macros. One notable example is [real-async-trait](https://crates.io/crates/real-async-trait).

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- None.

# Future possibilities
[future-possibilities]: #future-possibilities

### Naming return types

> We expect to introduce a mechanism for naming the result of `-> impl Trait` return types in a follow-up RFC.

Reviewer 3: ```suggestion
We expect to introduce a mechanism for naming the result of `-> impl Trait` return types in a follow-up RFC.
```

This sentences is ambiguous. At first I thought you meant "introducing a named associated type into the trait", but now I think you mean something like RTN? It'd be useful to reword it for clarity. In particular, I think what this means is "some way for generic code using a trait to place bounds on the return type of `-> impl Trait` functions"? (This is slightly different from being able to name the results of arbitrary `-> impl Trait` functions, e.g., if a function returns `-> (impl Trait, impl Trait)`.)

    }
}
```

## Scoping rules for generic parameters

We say a generic parameter is "in scope" for an `impl Trait` type if the actual revealed type is allowed to name that parameter. The scoping rules for return position `impl Trait` in traits are the same as [those for return position `impl Trait` generally][scoping]: All type and const parameters are considered in-scope, while lifetime parameters are only considered in-scope if they are mentioned in the `impl Trait` type directly.

Formally, given a trait method with a return type like `-> impl A + ... + Z` and an implementation of that trait, the hidden type for that implementation is allowed to reference:

* Concrete types, constant expressions, and `'static`
* Any generic type and const parameters in scope, including:
    * Type and const parameters on the impl
    * Explicit type and const parameters on the method
    * Implicit type parameters on the method (argument-position `impl Trait` types)
* Lifetime parameters that appear anywhere in the `impl A + ... + Z` type, including elided lifetimes

[scoping]: https://rust-lang.github.io/rfcs/1951-expand-impl-trait.html#scoping-for-type-and-lifetime-parameters

> Lifetime parameters not in scope may still be indirectly named by one of the type parameters in scope.

Author: The intent is to be consistent with inherent methods, but the implementation is not currently. I believe this is captured in the unresolved question around https://github.com/rust-lang/rust/issues/112194.

```rust
trait NewIntoIterator {
    type Item;
    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
}
```

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

*This section assumes familiarity with the [basic semantics of impl trait in return position][rpit].*

When you use `impl Trait` as the return type for a function within a trait definition or trait impl, the intent is the same: impls that implement this trait return "some type that implements `Trait`", and users of the trait can only rely on that.

<!--However, the desugaring to achieve that effect looks somewhat different than other cases of impl trait in return position. This is because we cannot desugar to a type alias in the surrounding module; we need to desugar to an associated type (effectively, a type alias in the trait).-->

Consider the following trait:

```rust
trait IntoIntIterator {
    fn into_int_iter(self) -> impl Iterator<Item = u32>;
}
```

The semantics of this are analogous to introducing a new associated type within the surrounding trait;

```rust

> trait IntoIntIterator { // desugared

Reviewer 5: Or `U32Iterator`
